File: android_app/src/main/java/me/timeto/app/ui/EventTemplateFormSheet.kt
package me.timeto.app.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import me.timeto.app.*
import me.timeto.shared.db.EventTemplateDb
import me.timeto.shared.vm.EventTemplateFormSheetVM

@Composable
fun EventTemplateFormSheet(
    layer: WrapperView.Layer,
    eventTemplate: EventTemplateDb?,
) {

    val (vm, state) = rememberVM(eventTemplate) {
        EventTemplateFormSheetVM(eventTemplate)
    }

    VStack(
        modifier = Modifier
            .fillMaxHeight()
            .background(c.sheetBg)
            .imePadding(),
    ) {

        val scrollState = rememberScrollState()

        Sheet__HeaderView(
            title = state.headerTitle,
            scrollState = scrollState,
            bgColor = c.sheetBg,
        )

        Column(
            modifier = Modifier
                .verticalScroll(state = scrollState)
                .padding(bottom = 20.dp)
                .weight(1f)
        ) {

            MyListView__ItemView(
                isFirst = true,
                isLast = true,
                modifier = Modifier
                    .padding(top = 12.dp)
            ) {

                MyListView__ItemView__TextInputView(
                    placeholder = "Text",
                    text = state.inputTextValue,
                    onTextChanged = { vm.setInputTextValue(it) },
                )
            }

            MyListView__Padding__SectionSection()

            MyListView__ItemView(
                isFirst = true,
                isLast = true,
            ) {
                MyListView__ItemView__ButtonView(
                    text = state.daytimeTitle,
                    withArrow = true,
                    rightView = {
                        MyListView__ItemView__ButtonView__RightText(
                            text = state.daytimeNote,
                            paddingEnd = 2.dp,
                            color = state.daytimeNoteColor?.toColor(),
                        )
                    }
                ) {
                    Sheet.show { layer ->
                        DaytimePickerSheet(
                            layer = layer,
                            title = state.daytimeTitle,
                            doneText = "Done",
                            daytimePickerUi = state.defDaytimePickerUi,
                            withRemove = true,
                            onPick = { daytimePickerUi ->
                                vm.setDaytime(daytimePickerUi)
                            },
                            onRemove = {
                                vm.setDaytime(null)
                            },
                        )
                    }
                }
            }

            MyListView__Padding__SectionSection()

            TextFeaturesTimerFormView(state.textFeatures) {
                vm.setTextFeatures(it)
            }

            MyListView__Padding__SectionSection()

            TextFeaturesTriggersFormView(state.textFeatures) {
                vm.setTextFeatures(it)
            }

            val eventTemplateDB = vm.eventTemplateDB
            if (eventTemplateDB != null) {

                MyListView__Padding__SectionSection()

                MyListView__ItemView(
                    isFirst = true,
                    isLast = true,
                ) {
                    MyListView__ItemView__ActionView(
                        text = state.deleteText,
                    ) {
                        vm.delete(eventTemplateDB) {
                            layer.close()
                        }
                    }
                }
            }
        }

        Sheet__BottomViewDefault(
            primaryText = state.doneText,
            primaryAction = {
                vm.save {
                    layer.close()
                }
            },
            secondaryText = "Cancel",
            secondaryAction = {
                layer.close()
            },
        )
    }
}

File: ios_app/iphone/EventTemplateFormSheet.swift
import SwiftUI
import shared

struct EventTemplateFormSheet: View {
    
    @Binding private var isPresented: Bool
    
    @State private var vm: EventTemplateFormSheetVm
    
    @EnvironmentObject private var nativeSheet: NativeSheet
    
    @State private var scroll = 0
    
    init(
        isPresented: Binding<Bool>,
        eventTemplateDb: EventTemplateDb?
    ) {
        _isPresented = isPresented
        _vm = State(initialValue: EventTemplateFormSheetVm(
            eventTemplateDB: eventTemplateDb
        ))
    }
    
    var body: some View {
        
        VMView(vm: vm, stack: .VStack()) { state in
            
            Sheet__HeaderView(
                title: state.headerTitle,
                scrollToHeader: scroll,
                bgColor: c.sheetBg
            )
            
            ScrollViewWithVListener(showsIndicators: false, vScroll: $scroll) {
                
                VStack {
                    
                    MyListView__ItemView(
                        isFirst: true,
                        isLast: true
                    ) {
                        
                        MyListView__ItemView__TextInputView(
                            text: state.inputTextValue,
                            placeholder: "Text",
                            isAutofocus: false,
                            onValueChanged: { newText in vm.setInputTextValue(text: newText) }
                        )
                    }
                    .padding(.top, 12)
                    
                    MyListView__Padding__SectionSection()
                    
                    MyListView__ItemView(
                        isFirst: true,
                        isLast: true
                    ) {
                        
                        MyListView__ItemView__ButtonView(
                            text: state.daytimeTitle,
                            withArrow: true,
                            rightView: AnyView(
                                MyListView__ItemView__ButtonView__RightText(
                                    text: state.daytimeNote,
                                    paddingEnd: 2,
                                    textColor: state.daytimeNoteColor?.toColor()
                                )
                            )
                        ) {
                            nativeSheet.show { isTimerPickerPresented in
                                DaytimePickerSheet(
                                    isPresented: isTimerPickerPresented,
                                    title: state.daytimeTitle,
                                    doneText: "Done",
                                    daytimeUi: state.defDaytimeUi,
                                    onPick: { daytimeUi in
                                        vm.setDaytime(daytimeUi: daytimeUi)
                                    },
                                    onRemove: {
                                        vm.setDaytime(daytimeUi: nil)
                                    }
                                )
                                .presentationDetents([.medium])
                                .presentationDragIndicator(.visible)
                            }
                        }
                    }
                    
                    MyListView__Padding__SectionSection()
                    
                    TextFeaturesTimerFormView(
                        textFeatures: state.textFeatures
                    ) { textFeatures in
                        vm.setTextFeatures(newTextFeatures: textFeatures)
                    }
                    
                    MyListView__Padding__SectionSection()
                    
                    TextFeaturesTriggersFormView(
                        textFeatures: state.textFeatures
                    ) { textFeatures in
                        vm.setTextFeatures(newTextFeatures: textFeatures)
                    }
                    
                    if let eventTemplate = vm.eventTemplateDB {
                        
                        MyListView__Padding__SectionSection()
                        
                        MyListView__ItemView(
                            isFirst: true,
                            isLast: true
                        ) {
                            MyListView__ItemView__ActionView(
                                text: state.deleteText
                            ) {
                                vm.delete(templateDB: eventTemplate) {
                                    isPresented = false
                                }
                            }
                        }
                    }
                }
            }
            
            Sheet__BottomViewDefault(
                primaryText: state.doneText,
                primaryAction: {
                    vm.save {
                        isPresented = false
                    }
                },
                secondaryText: "Cancel",
                secondaryAction: {
                    isPresented = false
                },
                topContent: { EmptyView() },
                startContent: { EmptyView() }
            )
        }
    }
}
